airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/InitialSnapshotHandler.java

```diff
@@ -0,0 +1,132 @@
+import io.airbyte.integrations.source.mongodb.internal.state.MongoDbStateManager;
+import io.airbyte.integrations.source.mongodb.internal.state.MongoDbStreamState;
+import io.airbyte.protocol.models.v0.AirbyteMessage;
+import io.airbyte.protocol.models.v0.CatalogHelpers;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.v0.SyncMode;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Retrieves iterators used for the initial snapshot
+ */
+public class InitialSnapshotHandler {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(InitialSnapshotHandler.class);
+
+  /**
+   * For each given stream configured as incremental sync it will output an iterator that will
+   * retrieve documents from the given database. Each iterator will start after the last checkpointed
+   * document, if any, or from the beginning of the stream otherwise.
+   */
+  public List<AutoCloseableIterator<AirbyteMessage>> getIterators(
+                                                                  final List<ConfiguredAirbyteStream> streams,
+                                                                  final MongoDbStateManager stateManager,
+                                                                  final MongoDatabase database,
+                                                                  final Instant emittedAt) {
+    return streams
+        .stream()
+        .peek(airbyteStream -> {
+          if (!airbyteStream.getSyncMode().equals(SyncMode.INCREMENTAL))
+            LOGGER.warn("Stream {} configured with unsupported sync mode: {}", airbyteStream.getStream().getName(), airbyteStream.getSyncMode());
+        })
+        .filter(airbyteStream -> airbyteStream.getSyncMode().equals(SyncMode.INCREMENTAL))
+        .map(airbyteStream -> {
+          final var collectionName = airbyteStream.getStream().getName();
+          final var collection = database.getCollection(collectionName);
+          final var fields = Projections.fields(Projections.include(CatalogHelpers.getTopLevelFieldNames(airbyteStream).stream().toList()));
+
+          final var idTypes = aggregateIdField(collection);
+          if (idTypes.size() != 1) {
+            throw new ConfigErrorException("The _id fields in a collection must be consistently typed.");
+          }
+
+          if (IdType.findByBsonType(idTypes.get(0)).isEmpty()) {
+            throw new ConfigErrorException("Only _id fields with the following types are currently supported: " + IdType.SUPPORTED);
+          }
+
+          // find the existing state, if there is one, for this stream
+          final Optional<MongoDbStreamState> existingState =
```

Comment:
What happens if the saved offset is no longer valid. Then we can no longer use this state. The state manager needs to be reset

Reply:
@subodh1810 I'm not sure that applies here?  This code is retrieving the initial snapshot checkpoint.  That data doesn't have to exist for us to perform the initial snapshot.  This code uses the `_id` value from the documents in the collection, not the information in the oplog.  If, for some reason, the user has deleted the document with the `_id` value that was last saved to the state (and any other documents beyond it), this query will still work.  It simply does a > "last checkpointed _id value".

Reply:
@subodh1810 and I discussed this and the code has been updated to reset the state manager if the offset state is invalid.